### PR TITLE
snap: add key management commands

### PR DESCRIPTION
**Pull request by cjwatson**
This adds four new hidden commands to snap, intended mainly for use by
snapcraft as an abstraction layer over GPG that knows how to deal with
SHA3-384 fingerprints: create-key, delete-key, export-key, and keys.

Specification, including rationale: https://docs.google.com/document/d/1Bkn9sLgXqkGsgV1eBWO935z5jZJiJQCL24muqw2zD7U

### CODE REVIEW

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by mvo5** [comment]
whitelist this please

File: cmd/snap/cmd_create_key_test.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+package main_test
+
+import (
+	. "gopkg.in/check.v1"
+
+	snap "github.com/snapcore/snapd/cmd/snap"
+)
+
+func (s *SnapSuite) TestCreateKeyInvalidCharacters(c *C) {
```
Comment (niemeyer) [inline cmd/snap/cmd_create_key_test.go]: No test of it succeeding?

File: cmd/snap/cmd_create_key.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+type cmdCreateKey struct {
+	Positional struct {
+		KeyName string `positional-arg-name:"<key-name>" description:"name of key to create; defaults to 'default'"`
+	} `positional-args:"true"`
+}
+
+func init() {
+	cmd := addCommand("create-key",
+		i18n.G("Create key"),
```
Comment (niemeyer) [inline cmd/snap/cmd_create_key.go]: That's implied in the command name. G("Create cryptographic key pair")?

File: cmd/snap/cmd_delete_key.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+type cmdDeleteKey struct {
+	Positional struct {
+		KeyName string `positional-arg-name:"<key-name>" description:"name of key to delete"`
+	} `positional-args:"true" required:"true"`
+}
+
+func init() {
+	cmd := addCommand("delete-key",
+		i18n.G("Delete key"),
```
Comment (niemeyer) [inline cmd/snap/cmd_delete_key.go]: `"Delete cryptographic key pair"`

File: cmd/snap/cmd_delete_key.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+		i18n.G("Delete an assertion-signing key."),
```
Comment (niemeyer) [inline cmd/snap/cmd_delete_key.go]: "Delete the local cryptographic key pair with the provided name."

File: cmd/snap/cmd_export_key.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+type cmdExportKey struct {
+	Positional struct {
+		KeyName string `positional-arg-name:"<key-name>" description:"name of key to export"`
+	} `positional-args:"true" required:"true"`
+}
+
+func init() {
+	cmd := addCommand("export-key",
+		i18n.G("Export key"),
```
Comment (niemeyer) [inline cmd/snap/cmd_export_key.go]: `"Export cryptographic key pair"`

Comment (niemeyer) [inline cmd/snap/cmd_export_key.go]: `"Export cryptographic public key"`

File: cmd/snap/cmd_export_key.go [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+		i18n.G("Export a public key in the format required by assertions."),
```
Comment (niemeyer) [inline cmd/snap/cmd_export_key.go]: `"Export a public key assertion that may be imported in other systems."`

Or something similar. It's not in the format required by assertions.. it _is_ an assertion itself.

File: cmd/snap/cmd_keys.go [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/jessevdk/go-flags"
+)
+
+type cmdKeys struct {
+	JSON bool `long:"json" description:"output results in JSON format"`
+}
+
+func init() {
+	cmd := addCommand("keys",
+		i18n.G("List keys"),
```
Comment (niemeyer) [inline cmd/snap/cmd_keys.go]: `List cryptographic keys`

File: cmd/snap/cmd_keys_test.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main_test
+
+//go:generate go-bindata -pkg main_test -nocompress -o ./cmd_keys_data_test.go test-data/
```
Comment (niemeyer) [inline cmd/snap/cmd_keys_test.go]: Why not just opening the actual file in test-data/ below?  All of the go-bindata and special handling of the magical test file can go away I believe.

File: asserts/gpgkeypairmgr.go [hunk 9]
```diff
@@ -23,15 +23,51 @@ import (
...
+		return "", err
+	}
+
+	gid, err := strconv.Atoi(real.Gid)
+	if err != nil {
+		return "", err
+	}
+
+	if homedir == "" {
+		homedir = os.Getenv("SNAP_GNUPGHOME")
```
Comment (niemeyer) [inline asserts/gpgkeypairmgr.go]: `SNAP_GNUPG_HOME` please.

File: asserts/gpgkeypairmgr.go [hunk 10]
```diff
@@ -23,15 +23,51 @@ import (
...
+	}
+	if homedir == "" {
+		homedir = filepath.Join(real.HomeDir, ".snap")
```
Comment (niemeyer) [inline asserts/gpgkeypairmgr.go]: Can we please move this under ~/.snap/gnupg/, so we don't risk any surprising conflicts.

**Comment by niemeyer** [comment]
Looks quite reasonable. Only a few minor points.

File: .travis.yml [hunk 11]
```diff
@@ -18,5 +18,6 @@ install:
   - sudo apt-get update -qq
   - sudo apt-get install -qq squashfs-tools
   - sudo apt-get install -qq gnupg
+  - sudo apt-get install -qq go-bindata
```
Comment (zyga) [inline .travis.yml]: Is this a new build dependency that should go in debian/control?

**Comment by pedronis** [comment]
test this please

**Comment by pedronis** [comment]
retest this please

**Comment by chipaca** [comment]
add to whitelist

**Comment by chipaca** [comment]
@mvo5 @pedronis for your reference, https://wiki.jenkins-ci.org/display/JENKINS/GitHub+pull+request+builder+plugin

File: asserts/gpgkeypairmgr.go [hunk 12]
```diff
@@ -23,15 +23,51 @@ import (
...
+	"os"
 	"os/exec"
+	"path/filepath"
+	"strconv"
 	"strings"
+
+	"github.com/snapcore/snapd/osutil"
 )
 
+func ensureGPGHomeDirectory(homedir string) (string, error) {
```
Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: wondering if we still need to support passing in a homedir here? given the strong preference for the default location (and we have the env var for tests if needed)

File: cmd/snap/cmd_create_key_test.go (hunk 1 above)
Comment (cjwatson) [inline cmd/snap/cmd_create_key_test.go]: The reason for the missing test here is that I haven't figured out how to deal with testing code that involves terminal.ReadPassword; "snap login" is already similarly untested, I assume for the same reason.  Any advice would be welcome.

File: asserts/gpgkeypairmgr.go [hunk 13]
```diff
@@ -23,15 +23,51 @@ import (
...
-		general = append([]string{"--homedir", homedir}, general...)
+	homedir, err := ensureGPGHomeDirectory(homedir)
+	if err != nil {
+		return nil, err
 	}
+
+	general := []string{"--homedir", homedir, "-q", "--no-auto-check-trustdb"}
 	allArgs := append(general, args...)
 
 	cmd := exec.Command("gpg", allArgs...)
```
Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: given that this is now our own keyring, should we switch to use the more modern gpg2, I remember trying with master and things worked out of the box, don't know about the new code here though?

File: asserts/gpgkeypairmgr.go [hunk 14]
```diff
@@ -183,3 +223,83 @@ func (gkm *GPGKeypairManager) sign(fingerprint string, content []byte) ([]byte,
...
 }
+
+type gpgKeypairInfo struct {
+	pubKey      PublicKey
+	fingerprint string
+}
+
+func (gkm *GPGKeypairManager) findByName(name string) (*gpgKeypairInfo, error) {
+	stop := errors.New("stop marker")
+	var hit gpgKeypairInfo
```
Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: just personal, but I would have made this one *gpgKeyPairInfo

File: cmd/snap/cmd_export_key.go (hunk 6 above)
Comment (cjwatson) [inline cmd/snap/cmd_export_key.go]: It's the body of an assertion, not the whole thing (which would require further information such as the account-id).  But I'll rephrase it to "Export a public key assertion body ..." etc.

File: asserts/gpgkeypairmgr.go [hunk 15]
```diff
@@ -183,3 +223,83 @@ func (gkm *GPGKeypairManager) sign(fingerprint string, content []byte) ([]byte,
...
+	return nil, fmt.Errorf("cannot find key named %q in GPG keyring", name)
+}
+
+var generateTemplate = `
+Key-Type: RSA
+Key-Length: 4096
+Subkey-Type: RSA
+Subkey-Length: 4096
+Name-Real: %s
+Preferences: SHA512
```
Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: I think for sanity we want to set the Creation-Date to v1FixedTimestamp, with that the gpg fingerprint of this and the one of the public key that goes into the account-key would agree,

it's not a must though

File: asserts/gpgkeypairmgr.go [hunk 16]
```diff
@@ -183,3 +223,83 @@ func (gkm *GPGKeypairManager) sign(fingerprint string, content []byte) ([]byte,
...
+// Generate creates a new key with the given passphrase and name.
+func (gkm *GPGKeypairManager) Generate(passphrase string, name string) error {
+	_, err := gkm.findByName(name)
+	if err == nil {
+		return fmt.Errorf("key named %q already exists in GPG keyring", name)
+	}
+	generateParams := fmt.Sprintf(generateTemplate, name)
+	if passphrase != "" {
+		generateParams += "Passphrase: " + passphrase + "\n"
+	}
```
Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: I think with gpg2 we could ask it with a %command or it would default to ask for a passphrase on its own even with --batch

File: cmd/snap/cmd_export_key.go (hunk 6 above)
Comment (pedronis) [inline cmd/snap/cmd_export_key.go]: no this is right, this not exporting an assertion (it cannot sign it) just what would go in its body, we could invent something like serial-request, so account-key-request if really wanted though

File: asserts/gpgkeypairmgr.go (hunk 13 above)
Comment (cjwatson) [inline asserts/gpgkeypairmgr.go]: This isn't a very strong preference, but I'm somewhat inclined to stick with gpg.  The OS direction is for gpg2 to become gpg once it's ready to be used as the default; this is the way it is in Debian unstable for instance, and possibly yakkety too though I haven't checked.  So for me, saying "gpg" is more like "system default GnuPG implementation" than "GnuPG 1".

File: asserts/gpgkeypairmgr.go (hunk 14 above)
Comment (cjwatson) [inline asserts/gpgkeypairmgr.go]: That seems kind of inconsistent with the existing GPGKeypairManager and indeed all the various KeypairManager interfaces.

File: asserts/gpgkeypairmgr.go [hunk 17]
```diff
@@ -100,7 +136,7 @@ func (gkm *GPGKeypairManager) retrieve(fpr string) (PrivateKey, error) {
 
 // Walk iterates over all the RSA private keys in the local GPG setup calling the provided callback until this returns an error
 // TODO: revisit exposing this
```
Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: it's clear now we have external use cases, this TODO can be removed

File: cmd/snap/cmd_create_key_test.go (hunk 1 above)
Comment (pedronis) [inline cmd/snap/cmd_create_key_test.go]: @cjwatson maybe a functional/spread test using  expect?  see tests/main/login maybe

File: asserts/gpgkeypairmgr.go (hunk 13 above)
Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: ok, sounds reasonable, should we try to run the suite with boths then? to be prepared?

Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: we probably would need to add tests to GPGKeypairMananger own tests (not just through the cmd_*_test) to achieve that, anyway these are not blockers for this branch

File: asserts/gpgkeypairmgr.go (hunk 14 above)
Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: @cjwatson I honestly don't understand what you are referring to exactly, anyway not a major point either way

Comment (cjwatson) [inline asserts/gpgkeypairmgr.go]: I mean that there is already lots of code using the term "KeypairManager", and so "KeypairInfo" should surely be capitalised consistently with that.  I don't much care which way round it goes, of course, so maybe a future branch could rename KeypairManager etc. to KeyPairManager?

Comment (pedronis) [inline asserts/gpgkeypairmgr.go]: @cjwatson sorry there's a typo in my suggestion, my point is having the var be a pointer (*), not changing the name, but I fatfingered a capital P in there

**Comment by elopio** [comment]
retest this please

File: cmd/snap/cmd_create_key_test.go (hunk 1 above)
Comment (cjwatson) [inline cmd/snap/cmd_create_key_test.go]: All right, I wrote a spread test and finally managed to get it to pass.

**Comment by cjwatson** [comment]
I think this is ready for another look now.

**Comment by niemeyer** [comment]
Thanks! The export-key fix is still pending, but may be done on a follow up.

Merging now.
